File: Solution Heap.cpp

#include <queue>
#include <vector>
using namespace std;

class Solution {
    
public:
    //C++20: ... (span<const int> piles, int numberOfOperationsToDo) ...
    int minStoneSum(const vector<int>& piles, int numberOfOperationsToDo) const {
        int numberOfStonesRemaining = 0;
        priority_queue<int> maxHeap;
        for (int pile : piles) {
            maxHeap.push(pile);
            numberOfStonesRemaining += pile;
        }

        while (numberOfOperationsToDo-- > 0 && maxHeap.top() > 0) {
            int currentPile = maxHeap.top();
            maxHeap.pop();
            int numberOfStonesToSubtract = currentPile / 2;
            numberOfStonesRemaining -= numberOfStonesToSubtract;
            currentPile -= numberOfStonesToSubtract;
            maxHeap.push(currentPile);
        }

        return numberOfStonesRemaining;
    }
};

File: Solution Heap.js

//const {PriorityQueue} = require('@datastructures-js/priority-queue');
/**
 * @param {number[]} piles
 * @param {number} numberOfOperationsToDo
 * @return {number}
 */
var minStoneSum = function (piles, numberOfOperationsToDo) {
    let numberOfStonesRemaining = 0;
    const maxHeap = new MaxPriorityQueue({compare: (x, y) => y - x});
    for (let pile of piles) {
        maxHeap.enqueue(pile);
        numberOfStonesRemaining += pile;
    }

    while (numberOfOperationsToDo-- > 0 && maxHeap.front() > 0) {
        let currentPile = maxHeap.dequeue();
        let numberOfStonesToSubtract = Math.floor(currentPile / 2);
        numberOfStonesRemaining -= numberOfStonesToSubtract;
        currentPile -= numberOfStonesToSubtract;
        maxHeap.enqueue(currentPile);
    }

    return numberOfStonesRemaining;
};
